src/main/java/com/aws/greengrass/mqttclient/spool/Spool.java

```diff
@@ -273,6 +273,10 @@ private void persistentQueueSync(Iterable<Long> diskQueueOfIds, CloudMessageSpoo
         int numMessages = 0;
         int queueOfMessageIdInitSize = queueOfMessageId.size();
         for (long currentId : diskQueueOfIds) {
+            // Check if Queue of message IDs already contains this ID
+            if (queueOfMessageId.contains(currentId)) {
```

Comment:
this doesn't seem threadsafe when there are other threads adding and removing messages at the same time.

Reply:
I did not find something particular to contains method, but looks like from java Docs all implementations of BlockingDequeue are threadsafe for all actions
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/BlockingDeque.html

Reply:
just because it is safe doesn't mean it is correct.

Reply:
Think about what happens if a message is either added or removed before or after your `contains` check. Does that make the check and subsequent actions incorrect?

Reply:
No, they won't matter as any add operation around that time would result in a new incremented message ID that we don't care about while performing this sync. As for the case that this code change covers, it doesn't matter.
There could be some weird edge cases resulting from a series of actions. I can go over them separately...

Reply:
I'd like you to think about a way that we can potentially delete the queue. It will be taking up a lot of memory and we probably can find a better way

Reply:
Got it, yes agreed